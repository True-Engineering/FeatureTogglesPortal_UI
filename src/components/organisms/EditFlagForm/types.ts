import type {
  IFlag,
  IFlagTypeProp,
  IFlippingStrategyTypeProp,
} from '../../../types';

export type IFlagType = IFlagTypeProp | 'noType';

export type IFlippingStrategyType = IFlippingStrategyTypeProp | 'NoStrategy';

export interface IEditableFlag extends Omit<IFlag, 'type'> {
  type?: IFlagType;
}
