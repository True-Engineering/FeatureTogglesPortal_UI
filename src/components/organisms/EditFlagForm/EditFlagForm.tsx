import React, { FC, useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import clsx from 'clsx';
import {
  Button,
  Icon,
  Input,
  Select,
  SmartInput,
} from '@true-engineering/true-react-common-ui-kit';
import {
  INestedErrorsResult,
  INestedValidationConfig,
  validateNestedForm,
} from '@true-engineering/true-react-common-validator';
import { AutocompleteInput, Confirmation, ContextPopup } from '../../atoms';
import { useFeatureTogglesTheme } from '../../../hooks';
import {
  getErrorMessageFunction,
  hasAnyErrorsInside,
  requiredValidator,
  hasError,
  getChangeFormHandler,
  getBlurFormHandler,
} from '../../../utils';
import { FLAG_TYPES, IFlag } from '../../../types';
import type { IEditableFlag, IFlagType } from './types';
import { formStyles, iconButtonTweakStyles } from '../../commonStyles';
import { deleteConfirmTweakStyles } from './EditFlagForm.styles';

const EMPTY_FLAG: IFlag = {
  name: '',
  description: '',
  sprint: undefined,
  group: '',
  tag: undefined,
  type: undefined,
  environments: [],
};

const EMPTY_TOUCHED: Record<keyof IFlag, boolean> = {
  name: false,
  description: false,
  sprint: false,
  group: false,
  tag: false,
  type: false,
  environments: false,
};

const flagTypes: IFlagType[] = ['noType', ...FLAG_TYPES];

interface IEditFlagFormProps {
  initialFlag?: IFlag;
  groupsList?: string[];
  tagsList?: string[];
  isNewFlag?: boolean;
  onSaveChanges: (editedFlag: IFlag) => void;
  onCancel: () => void;
  onDelete?: (editedFlag: IFlag) => void;
}

export const editFlagValidationConfig: INestedValidationConfig<IEditableFlag> = {
  name: { ...requiredValidator },
};

export const convertFlag = (flag: IEditableFlag): IFlag => ({
  ...flag,
  type: flag.type === 'noType' ? undefined : flag.type,
});

const EditFlagForm: FC<IEditFlagFormProps> = ({
  initialFlag,
  groupsList,
  tagsList,
  isNewFlag = false,
  onSaveChanges,
  onCancel,
  onDelete,
}) => {
  const { classes } = useFeatureTogglesTheme('formStyles', formStyles, {});
  const { t } = useTranslation();
  const getErrorMessage = useMemo(() => getErrorMessageFunction(t), [t]);

  const [editableFlag, setEditableFlag] = useState<IEditableFlag>(
    initialFlag ?? EMPTY_FLAG,
  );
  const [touched, setTouched] = useState(EMPTY_TOUCHED);

  const [errors, setErrors] = useState<INestedErrorsResult<IEditableFlag>>(
    validateNestedForm(editableFlag, editFlagValidationConfig),
  );

  useEffect(() => {
    const newErrors = validateNestedForm(
      editableFlag,
      editFlagValidationConfig,
    );
    setErrors(newErrors);
  }, [editableFlag]);

  const handleChange = getChangeFormHandler(setEditableFlag);

  const handleBlur = getBlurFormHandler(setTouched);

  const handleSubmit = (event: React.SyntheticEvent) => {
    event.preventDefault();

    if (hasAnyErrorsInside(errors)) {
      return;
    }

    onSaveChanges(convertFlag(editableFlag));
  };

  return (
    <form className={classes.root} onSubmit={handleSubmit}>
      <div className={classes.title}>{t('featureFlag')}</div>
      <div
        className={clsx(classes.input, classes.shortInput, classes.inputGroup)}
      >
        <Input
          label={t('sprint')}
          value={editableFlag.sprint}
          onChange={handleChange('sprint')}
          isInvalid={hasError(errors.sprint) && touched.sprint}
          errorMessage={getErrorMessage(errors?.sprint)}
          onBlur={handleBlur('sprint')}
        />
      </div>
      <div className={clsx(classes.inputGroup, classes.input)}>
        <AutocompleteInput
          label={t('group')}
          value={editableFlag.group}
          onChange={handleChange('group')}
          errorMessage={getErrorMessage(errors?.group)}
          isInvalid={hasError(errors.group) && touched.group}
          onBlur={handleBlur('group')}
          options={groupsList}
        />
      </div>
      <div className={clsx(classes.input, classes.inputGroup)}>
        <Input
          label={t('task')}
          value={editableFlag.description}
          onChange={handleChange('description')}
          errorMessage={getErrorMessage(errors?.description)}
          isInvalid={hasError(errors.description) && touched.description}
          onBlur={handleBlur('description')}
        />
      </div>
      <div className={clsx(classes.inputGroup, classes.input)}>
        <SmartInput
          label={t('featureFlag')}
          smartType="default"
          regExp={/^[a-zA-Z0-9._-]*$/i}
          value={editableFlag.name}
          onChange={isNewFlag ? handleChange('name') : () => undefined}
          isDisabled={!isNewFlag}
          isRequired
          errorMessage={getErrorMessage(errors?.name)}
          isInvalid={hasError(errors.name) && touched.name}
          onBlur={handleBlur('name')}
        />
      </div>
      <div className={classes.inputGroup}>
        <div className={clsx(classes.shortInput)}>
          <Select
            options={flagTypes}
            label={t('type')}
            value={editableFlag.type ?? 'noType'}
            onChange={handleChange('type')}
            errorMessage={getErrorMessage(errors?.type)}
            isInvalid={hasError(errors.type) && touched.type}
            onBlur={handleBlur('type')}
            noMatchesLabel={t('noType')}
            placeholder={t('type')}
            convertValueToString={t}
            hasFloatingLabel
          />
        </div>
        <div className={clsx(classes.shortInput, classes.input)}>
          <AutocompleteInput
            label={t('tag')}
            value={editableFlag.tag}
            onChange={handleChange('tag')}
            errorMessage={getErrorMessage(errors?.tag)}
            isInvalid={hasError(errors.tag) && touched.tag}
            onBlur={handleBlur('tag')}
            options={tagsList}
          />
        </div>
      </div>
      <div className={classes.buttonsContainer}>
        {!isNewFlag && onDelete !== undefined && (
          <div className={clsx(classes.button, classes.buttonFloatLeft)}>
            <ContextPopup
              tweakStyles={deleteConfirmTweakStyles}
              popup={({ onClose }) => (
                <Confirmation
                  onConfirm={() => {
                    onClose();
                    onDelete(convertFlag(editableFlag));
                  }}
                  onCancel={() => onClose()}
                  message={t('confirmToDeleteFlag')}
                  buttonView="destructive"
                  buttonText={t('delete')}
                />
              )}
            >
              {({ isOpen }) => (
                <Button
                  size="xl"
                  isActive={isOpen}
                  view="outline"
                  tweakStyles={iconButtonTweakStyles}
                >
                  <div className={classes.trashIcon}>
                    <Icon type="trash-can" />
                  </div>
                </Button>
              )}
            </ContextPopup>
          </div>
        )}
        <div className={classes.button}>
          <Button view="secondary" size="xl" onClick={onCancel}>
            {t('cancel')}
          </Button>
        </div>
        <div className={classes.button}>
          <Button
            view="primary"
            size="xl"
            isFullWidth
            type="submit"
            isDisabled={hasAnyErrorsInside(errors)}
          >
            {t('save')}
          </Button>
        </div>
      </div>
    </form>
  );
};

export default EditFlagForm;
