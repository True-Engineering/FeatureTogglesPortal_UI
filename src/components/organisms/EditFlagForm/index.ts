export { default as EditFlagForm } from './EditFlagForm';
export * from './EditFlagForm';

export type { EditFlagFormStyles } from './EditFlagForm.styles';
