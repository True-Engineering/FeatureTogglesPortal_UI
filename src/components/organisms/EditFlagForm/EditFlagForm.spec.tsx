/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent, waitFor } from '@testing-library/react';
import EditFlagForm from './EditFlagForm';

const initialFlag = {
  name: 'flag',
  description: 'text',
  environments: [],
  sprint: 'sprint 1',
};

describe('EditFlagForm корректно отрабатывает сценарии', () => {
  test('Рендер на мок данных', async () => {
    const { container } = render(
      <EditFlagForm isNewFlag onSaveChanges={() => {}} onCancel={() => {}} />,
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  test('Сохраняет форму по клике на кнопку ', async () => {
    const onSaveChanges = jest.fn();
    const onCancel = jest.fn();
    const { getByText } = render(
      <EditFlagForm
        initialFlag={initialFlag}
        onSaveChanges={onSaveChanges}
        onCancel={onCancel}
      />,
    );

    fireEvent.click(getByText('Сохранить'));

    await waitFor(() => {
      expect(onSaveChanges).toBeCalledTimes(1);
      expect(onCancel).toBeCalledTimes(0);
    });
  });

  test('Отменяет сохраненение формы по клике на кнопку', async () => {
    const onSaveChanges = jest.fn();
    const onCancel = jest.fn();
    const { getByText } = render(
      <EditFlagForm
        initialFlag={initialFlag}
        onSaveChanges={onSaveChanges}
        onCancel={onCancel}
      />,
    );

    fireEvent.click(getByText('Отмена'));
    await waitFor(() => {
      expect(onSaveChanges).toBeCalledTimes(0);
      expect(onCancel).toBeCalledTimes(1);
    });
  });

  test('Не сохраняет форму, если не задано имя', async () => {
    const onSaveChanges = jest.fn();
    const onCancel = jest.fn();
    const { getByText, container } = render(
      <EditFlagForm
        isNewFlag
        onSaveChanges={onSaveChanges}
        onCancel={onCancel}
      />,
    );

    fireEvent.click(getByText('Сохранить'));

    expect(container).toMatchSnapshot();
    await waitFor(() => {
      expect(onSaveChanges).toBeCalledTimes(0);
      expect(onCancel).toBeCalledTimes(0);
    });
  });
});
