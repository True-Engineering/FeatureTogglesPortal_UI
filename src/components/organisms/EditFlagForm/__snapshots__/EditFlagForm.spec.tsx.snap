// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFlagForm корректно отрабатывает сценарии Не сохраняет форму, если не задано имя 1`] = `
<div>
  <form
    class="root"
  >
    <div
      class="title"
    >
      Feature Flag
    </div>
    <div
      class="input shortInput inputGroup"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Спринт
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup input"
    >
      <div
        class="root"
      >
        <div
          class="root"
        >
          <div
            class="inputWrapper"
          >
            <input
              class="input withFloatingLabel"
              type="text"
              value=""
            />
            <span
              class="label floating"
            >
              Группа
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="input inputGroup"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Задача
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup input"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper required"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Feature Flag
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup"
    >
      <div
        class="shortInput"
      >
        <div
          class="root"
        >
          <div
            class="inputWrapper"
          >
            <div
              class="root"
            >
              <div
                class="inputWrapper"
              >
                <input
                  class="input withFloatingLabel"
                  readonly=""
                  type="text"
                  value="Без типа"
                />
                <span
                  class="label activeLabel floating"
                >
                  Тип
                </span>
              </div>
            </div>
            <div
              class="arrow"
            >
              <div
                class="root"
              >
                <svg
                  fill="none"
                  height="100%"
                  viewBox="0 0 20 20"
                  width="100%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10.0029 11.6053L5.71004 7.31242C5.31951 6.9219 4.68635 6.9219 4.29582 7.31242C3.9053 7.70295 3.9053 8.33611 4.29582 8.72664L9.29582 13.7266C9.68635 14.1172 10.3195 14.1172 10.71 13.7266L15.71 8.72664C16.1006 8.33611 16.1006 7.70295 15.71 7.31242C15.3195 6.9219 14.6863 6.9219 14.2958 7.31242L10.0029 11.6053Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    stroke="none"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="shortInput input"
      >
        <div
          class="root"
        >
          <div
            class="root"
          >
            <div
              class="inputWrapper"
            >
              <input
                class="input withFloatingLabel"
                type="text"
                value=""
              />
              <span
                class="label floating"
              >
                Тэг
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="buttonsContainer"
    >
      <div
        class="button"
      >
        <button
          class="root xl cancel"
          type="button"
        >
          <span
            class="content"
          >
            Отмена
          </span>
        </button>
      </div>
      <div
        class="button"
      >
        <button
          class="root xl main disabled fullWidth"
          disabled=""
          type="submit"
        >
          <span
            class="content"
          >
            Сохранить
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`EditFlagForm корректно отрабатывает сценарии Рендер на мок данных 1`] = `
<div>
  <form
    class="root"
  >
    <div
      class="title"
    >
      Feature Flag
    </div>
    <div
      class="input shortInput inputGroup"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Спринт
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup input"
    >
      <div
        class="root"
      >
        <div
          class="root"
        >
          <div
            class="inputWrapper"
          >
            <input
              class="input withFloatingLabel"
              type="text"
              value=""
            />
            <span
              class="label floating"
            >
              Группа
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="input inputGroup"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Задача
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup input"
    >
      <div
        class="root"
      >
        <div
          class="inputWrapper required"
        >
          <input
            class="input withFloatingLabel"
            type="text"
            value=""
          />
          <span
            class="label floating"
          >
            Feature Flag
          </span>
        </div>
      </div>
    </div>
    <div
      class="inputGroup"
    >
      <div
        class="shortInput"
      >
        <div
          class="root"
        >
          <div
            class="inputWrapper"
          >
            <div
              class="root"
            >
              <div
                class="inputWrapper"
              >
                <input
                  class="input withFloatingLabel"
                  readonly=""
                  type="text"
                  value="Без типа"
                />
                <span
                  class="label activeLabel floating"
                >
                  Тип
                </span>
              </div>
            </div>
            <div
              class="arrow"
            >
              <div
                class="root"
              >
                <svg
                  fill="none"
                  height="100%"
                  viewBox="0 0 20 20"
                  width="100%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10.0029 11.6053L5.71004 7.31242C5.31951 6.9219 4.68635 6.9219 4.29582 7.31242C3.9053 7.70295 3.9053 8.33611 4.29582 8.72664L9.29582 13.7266C9.68635 14.1172 10.3195 14.1172 10.71 13.7266L15.71 8.72664C16.1006 8.33611 16.1006 7.70295 15.71 7.31242C15.3195 6.9219 14.6863 6.9219 14.2958 7.31242L10.0029 11.6053Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    stroke="none"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="shortInput input"
      >
        <div
          class="root"
        >
          <div
            class="root"
          >
            <div
              class="inputWrapper"
            >
              <input
                class="input withFloatingLabel"
                type="text"
                value=""
              />
              <span
                class="label floating"
              >
                Тэг
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="buttonsContainer"
    >
      <div
        class="button"
      >
        <button
          class="root xl cancel"
          type="button"
        >
          <span
            class="content"
          >
            Отмена
          </span>
        </button>
      </div>
      <div
        class="button"
      >
        <button
          class="root xl main disabled fullWidth"
          disabled=""
          type="submit"
        >
          <span
            class="content"
          >
            Сохранить
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;
